### PR TITLE
Add GEM unpacker in L1Repack FullMC and temporary load of GEM geometry to hltGetConfiguration

### PR DESCRIPTION
**Pull request by srimanob**
#### PR description:

Move to https://github.com/cms-sw/cmssw/pull/34788

This PR is to fix two issues when L1REPACK is used for Run3. The error messages w/o this PR is mentioned at the end of the PR description.
(1) no GEM unpacker defined in SimL1EmulatorRepack. It is solved in this PR.
(2) when hltGetConfiguration is used, GEM geometry is not loaded properly. This PR introduced a temporary fix.

#### PR validation:
The RAW input can be picked from `/afs/cern.ch/user/s/srimanob/public/ForGEML1T/step2.root`. It is a RAW file produced from CMSSW_12_0_0_pre4 with wf 11634.21.

To test (1), the following cmsDriver run fine.
`cmsDriver.py --process reHLT -s L1REPACK:FullMC,HLT:@relval2021 --conditions auto:phase1_2021_realistic --datatier GEN-SIM-RAW -n -1 --eventcontent RAWSIM --geometry DB:Extended --era Run3 --filein file:step2.root --fileout file:step3_reHLT.root --python reHLT_ProdLike_2021.py --no_exec`

To test (2), the following config can run fine.
`hltGetConfiguration /dev/CMSSW_12_0_0/GRun/V2 --process reHLT --globaltag auto:phase1_2021_realistic  --mc --unprescale --paths HLTriggerFirstPath,HLT_IsoMu24_v13,HLT_Mu50_v13,HLTriggerFinalPath --output minimal --input file:step2.root --l1-emulator FullMC --customise HLTrigger/Configuration/customizeHLTforCMSSW.customiseForRun3GEMGeometry > test_reHLT.py`

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
Not a backport, and no need of backport

#### Note
Issue of (1): solved in this PR
```
----- Begin Fatal Exception 05-Aug-2021 05:33:59 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'HLTAnalyzerEndpath'
   [2] Prefetching for module L1TRawToDigi/'hltGtStage2Digis'
   [3] Prefetching for module RawDataCollectorByLabel/'rawDataCollector'
   [4] Prefetching for module L1TDigiToRaw/'packGmtStage2'
   [5] Prefetching for module L1TMuonOverlapTrackProducer/'simOmtfDigis'
   [6] Prefetching for module CSCTriggerPrimitivesProducer/'simCscTriggerPrimitiveDigis'
   [7] Prefetching for module GEMPadDigiClusterProducer/'simMuonGEMPadDigiClusters'
   [8] Calling method for module GEMPadDigiProducer/'simMuonGEMPadDigis'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: MuonDigiCollection<GEMDetId,GEMDigi>
Looking for module label: simMuonGEMDigis
Looking for productInstanceName: 
```

Issue of (2): temporary solution is provided for now
```
----- Begin Fatal Exception 05-Aug-2021 05:59:31 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  stream begin Run run: 1 stream: 1
   [1] Calling method for module GEMPadDigiClusterProducer/'simMuonGEMPadDigiClusters'
Exception Message:
No data of type "GEMGeometry" with label "" in record "MuonGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
----- Begin Fatal Exception 05-Aug-2021 05:59:31 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  stream begin Run run: 1 stream: 3
   [1] Calling method for module GEMPadDigiProducer/'simMuonGEMPadDigis'
Exception Message:
No data of type "GEMGeometry" with label "" in record "MuonGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
```

### CODE REVIEW

**Comment by srimanob** [comment]
FYI @dildick @jshlee @rekovic

**Comment by srimanob** [comment]
Hi @Martin-Grunewald @fwyzard 
Do you have a hint of what can make the difference between the following command? Thanks in advance.

(1)
`cmsDriver.py --process reHLT -s L1REPACK:FullMC,HLT:@relval2021 --conditions auto:phase1_2021_realistic --datatier GEN-SIM-RAW -n -1 --eventcontent RAWSIM --geometry DB:Extended --era Run3 --filein file:step2.root --fileout file:step3_reHLT.root --python reHLT_ProdLike_2021.py --no_exec`

(2)
`hltGetConfiguration /dev/CMSSW_12_0_0/GRun/V2 --process reHLT --globaltag auto:phase1_2021_realistic --mc --unprescale --paths HLTriggerFirstPath,HLT_IsoMu24_v13,HLT_Mu50_v13,HLTriggerFinalPath --output minimal --input file:step2.root --l1-emulator FullMC > test_reHLT.py`
I assume Run3 era in (2) is defined in 
https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/CustomConfigs.py#L123-L127
already. I also try to add `--eras` but it does not help.

**Comment by dildick** [comment]
It seems that `hltGetConfiguration` is loading the wrong geometry.

**Comment by srimanob** [comment]
Hi @dildick 
If I add (*) to the hlt config, thing will run properly. It seems we have to find the place where this kind of geometry config is dumped by `hltGetConfiguration`. I am looking around in 
https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/Tools/confdb.py

```
process.GEMGeometryESModule = cms.ESProducer( "GEMGeometryESModule",
  fromDD4Hep = cms.bool( False ),
  appendToDataLabel = cms.string( "" ),
  fromDDD = cms.bool( False )
)
```

**Comment by perrotta** [comment]
> Hi @dildick
> If I add (*) to the hlt config, thing will run properly. It seems we have to find the place where this kind of geometry config is dumped by `hltGetConfiguration`. I am looking around in
> https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/Tools/confdb.py
> 
> ```
> process.GEMGeometryESModule = cms.ESProducer( "GEMGeometryESModule",
>   fromDD4Hep = cms.bool( False ),
>   appendToDataLabel = cms.string( "" ),
>   fromDDD = cms.bool( False )
> )
> ```

I think @Martin-Grunewald can suggest how to best fix it

**Comment by fwyzard** [comment]
For consistency with the other muon ESProducer modules, a solution could be
  - add `GEMGeometryESModule` to the HLT menu in ConfDB, based on the snippet [provided by Phat](https://github.com/cms-sw/cmssw/pull/34785#issuecomment-893194162)
  - update [HLTrigger/Configuration/python/Tools/confdb.py](https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/Tools/confdb.py) to remove it from the `fragment` dumps, in order to reuse from the offline event setup
  - regenerate the HLT dumps

**Comment by Martin-Grunewald** [comment]
Until then add a ```process.load``` of the GEMGeometryESModule cfi file in customizeHLTforCMSSW.... (process.load should not create a duplicate but overwrite if it is already there)

**Comment by srimanob** [comment]
Thanks very much @fwyzard @Martin-Grunewald 

I've provided the temporary solution as suggested. 

To add GEMGeometryESModule to the HLT menu in ConfDB, should I / GEM contact do it? I've never done it before. Thanks.

**Comment by srimanob** [comment]
Please test

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -134,6 +134,18 @@ def customiseFor34120(process):
     process.load("Geometry.TrackerGeometryBuilder.TrackerAdditionalParametersPerDet_cfi")
     return process
 
+#temporary solution to add GEM geometry for hltGetConfiguration
+def customiseForRun3GEMGeometry(process):
```
[the remark this answers is not in the document]
Comment (Martin-Grunewald) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:138]: You also need to call this function, at the bottom of the file!
```
      process = customiseForRun3GEMGeometry34785(process)

[the remark this answers is not in the document]
Comment (srimanob) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: Ah understand, Thanks. 
Added.

I tested with `--customise HLTrigger/Configuration/customizeHLTforCMSSW.customiseForRun3GEMGeometry` before, so I did not add it.

**Comment by Martin-Grunewald** [comment]
There is now a merge conflict in HLTrigger/Configuration/python/customizeHLTforCMSSW.py as that file has changed recently where you call the function. Please rebase...

**Comment by srimanob** [comment]
I mess up with rebase. Close and make a clean one instead.

**Pull request closed without merging**